cmd/rebasehelpers/util/git.go

```diff
@@ -0,0 +1,164 @@
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var UpstreamSummaryPattern = regexp.MustCompile(`UPSTREAM: (revert: [a-f0-9]{7,}: )?(([\w\.-]+\/[\w-]+): )?([0-9]{4,}:|<carry>:|<drop>:)`)
+
+type Commit struct {
+	Sha     string
+	Summary string
+	Files   []File
+}
+
+func (c Commit) DeclaresUpstreamChange() bool {
+	return strings.HasPrefix(strings.ToLower(c.Summary), "upstream")
+}
+
+func (c Commit) MatchesUpstreamSummaryPattern() bool {
+	return UpstreamSummaryPattern.MatchString(c.Summary)
+}
+
+func (c Commit) DeclaredUpstreamRepo() (string, error) {
+	if !c.DeclaresUpstreamChange() {
+		return "", fmt.Errorf("commit declares no upstream changes")
+	}
+	if !c.MatchesUpstreamSummaryPattern() {
+		return "", fmt.Errorf("commit doesn't match the upstream commit summary pattern")
+	}
+	groups := UpstreamSummaryPattern.FindStringSubmatch(c.Summary)
+	repo := groups[3]
+	if len(repo) == 0 {
+		repo = "k8s.io/kubernetes"
+	}
+	return repo, nil
+}
+
+func (c Commit) HasGodepsChanges() bool {
+	for _, file := range c.Files {
+		if file.HasGodepsChanges() {
+			return true
+		}
+	}
+	return false
+}
+
+func (c Commit) HasNonGodepsChanges() bool {
+	for _, file := range c.Files {
+		if !file.HasGodepsChanges() {
+			return true
+		}
+	}
+	return false
+}
+
+func (c Commit) GodepsReposChanged() ([]string, error) {
+	repos := map[string]struct{}{}
+	for _, file := range c.Files {
+		if !file.HasGodepsChanges() {
+			continue
+		}
+		repo, err := file.GodepsRepoChanged()
+		if err != nil {
+			return nil, fmt.Errorf("problem with file %q in commit %s: %s", file, c.Sha, err)
+		}
+		repos[repo] = struct{}{}
+	}
+	changed := []string{}
+	for repo := range repos {
+		changed = append(changed, repo)
+	}
+	return changed, nil
+}
+
+type File string
+
+func (f File) HasGodepsChanges() bool {
+	return strings.HasPrefix(string(f), "Godeps")
+}
+
+func (f File) GodepsRepoChanged() (string, error) {
+	if !f.HasGodepsChanges() {
+		return "", fmt.Errorf("file doesn't appear to be a Godeps change")
+	}
+	// Find the _workspace path segment index.
+	workspaceIdx := -1
+	parts := strings.Split(string(f), string(os.PathSeparator))
+	for i, part := range parts {
+		if part == "_workspace" {
+			workspaceIdx = i
+			break
+		}
+	}
+	// Godeps path struture assumption: Godeps/_workspace/src/repo/name
```

Comment:
I believe I've fixed support for this, but the code and regex are now a little more complicated and we have to maintain a map of supported hosts -> expected segment count. Works, but very ugly.